### PR TITLE
Drivers: Expose jinja template parameters as static constexpr.

### PR DESCRIPTION
**Pull request by strongly-typed**
A static_assert in a driver can be used to check if e.g. a peripheral
provides a buffer of adequate size.

### CODE REVIEW

**Comment by ekiwi** [comment]
Looks reasonable, but can you change the implementations in the `*.cpp`/`*_impl.hpp` files to use the new `c++` constants instead of hard coding the `jinja2` template parameter?
This way the template parameter would be only used in one place and this might prevent some accidental mistakes when changing things.

**Comment by ekiwi** [comment]
Names of scoped constants should be written in `CamelCase`.

From the [coding conventions](https://github.com/roboterclubaachen/xpcc/blob/3292de0f26eb0cd1bf126bccd9827a552b20fe47/doc/coding_convention.md):
> Scoped constants, enums, structs must be Capitalized and not uppercased.

**Comment by ekiwi** [comment]
I think the doc string is redundant and only increases the maintenance burden:
> /// Expose jinja template parameters to be checked by e.g. drivers or application

This does not provide any useful information. Also you only document the first constant whereas the second constant in a class won't be associated with your doc string in the generated `doxygen` documentation.

File: src/xpcc/architecture/platform/driver/can/stm32/can.hpp.in [hunk 1]
```diff
@@ -77,6 +77,10 @@ public:
 	static const TypeId::Can{{ id }}Rx Rx;
 	/// TypeId used to connect GPIO pins to this peripheral's tx.
 	static const TypeId::Can{{ id }}Tx Tx;
+public:
+	/// Expose jinja template parameters to be checked by e.g. drivers or application
+	static constexpr uint8_t rxBufferSize = parameters.rx_buffer;
+	static constexpr uint8_t txBufferSize = parameters.tx_buffer;
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/can/stm32/can.hpp.in]: Missing braces

**Comment by salkinium** [comment]
This is a good idea, but the check should happen _before_ or latest _during_ template generation. Unfortunately the current build system doesn’t allow this, but modm does. If you don’t require these checks right now, I’d prefer not to add them, since I’ll have to remove or port them to modm. Do you have a specific issue you need to solve?

File: src/xpcc/architecture/platform/driver/can/stm32/can.hpp.in (hunk 1 above)
Comment (strongly-typed) [reply src/xpcc/architecture/platform/driver/can/stm32/can.hpp.in]: This was a test for the CI ;-)
Travis caught this, CircleCI's integration to GitHub does not yet run smooth. I'll have a look.

**Comment by strongly-typed** [comment]
Two use cases at the moment:
 * OneWire over UART. The OneWireMaster driver should be able to check if the provided UART has a buffer of appropriate size.
 * ROSserial integration into xpcc. The ROSserial hardware abstraction requires the hardware to provide a buffer of at least 512 bytes. It does not check if the UART is free to send enough bytes. Yes, this implementation should be improved, but the quick fix is to provide a buffer of appropriate size.

Checking during template generation is not possible because a UART with 64 bytes buffer on its own is totally fine. It only may become an issue if you want to use it for e.g. ROSserial. Would you know in modm at template generation that it will get connected?

**Comment by salkinium** [comment]
> Checking during template generation is not possible because a UART with 64 bytes buffer on its own is totally fine.

Ah, I misunderstood this then. Yes, that makes way more sense now and it's awesome! 

> Would you know in modm at template generation that it will get connected?

No, obviously not. There is a bit of ambiguity between what checks should happen at lbuild time and what at compile time. I'm still wrapping my head around that.

Can you update the interfaces and their documentation too? We need a common naming convention for these values. In the far future (when C++ concepts are standardized), we'll actually enforce these interfaces at compile time.

**Comment by rleh** [comment]
Some buffer sizes are (signed) `int`, some are `uint8_t`. Is there any reason for that?

**Comment by strongly-typed** [comment]
> files to use the new c++ constants instead of hard coding the jinja2 template parameter

I don't know how.

     %% if parameters.rx_buffer > 0
     static xpcc::atomic::Queue<xpcc::can::Message, {{ parameters.rx_buffer }}> rxQueue;
     %% endif

Because Jinja is used to create different code depending on the parameter. I can't use C++ `if` here.

**Review by salkinium — COMMENTED** [review]
All sizes should be `size_t`, otherwise they'll overflow when someone™ changes the parameter `max` values.

**Review by salkinium — APPROVED** [review]
Nice!

**Pull request merged**